sql/hive/src/test/scala/org/apache/spark/sql/hive/security/HiveHadoopDelegationTokenManagerSuite.scala

```diff
@@ -0,0 +1,106 @@
+
+import org.apache.commons.io.IOUtils
+import org.apache.hadoop.conf.Configuration
+
+import org.apache.spark.{SparkConf, SparkFunSuite}
+import org.apache.spark.deploy.security.HadoopDelegationTokenManager
+import org.apache.spark.util.Utils
+
+class HiveHadoopDelegationTokenManagerSuite extends SparkFunSuite {
+  private val hadoopConf = new Configuration()
+
+  test("default configuration") {
+    val manager = new HadoopDelegationTokenManager(new SparkConf(false), hadoopConf, null)
+    assert(manager.isProviderLoaded("hive"))
+  }
+
+  test("disable hive credential provider") {
+    val sparkConf = new SparkConf(false).set("spark.security.credentials.hive.enabled", "false")
+    val manager = new HadoopDelegationTokenManager(sparkConf, hadoopConf, null)
+    assert(!manager.isProviderLoaded("hive"))
+  }
+
+  test("using deprecated configurations") {
+    val sparkConf = new SparkConf(false)
+      .set("spark.yarn.security.credentials.hive.enabled", "false")
+    val manager = new HadoopDelegationTokenManager(sparkConf, hadoopConf, null)
+    assert(!manager.isProviderLoaded("hive"))
+  }
+
+  test("SPARK-23209: obtain tokens when Hive classes are not available") {
+    // This test needs a custom class loader to hide Hive classes which are in the classpath.
+    // Because the manager code loads the Hive provider directly instead of using reflection, we
+    // need to drive the test through the custom class loader so a new copy that cannot find
+    // Hive classes is loaded.
+    val currentLoader = Thread.currentThread().getContextClassLoader()
+    val noHive = new ClassLoader() {
+      override def loadClass(name: String, resolve: Boolean): Class[_] = {
+        if (name.startsWith("org.apache.hive") || name.startsWith("org.apache.hadoop.hive")) {
+          throw new ClassNotFoundException(name)
+        }
+
+        val prefixBlacklist = Seq("java", "scala", "com.sun.", "sun.")
+        if (prefixBlacklist.exists(name.startsWith(_))) {
+          return currentLoader.loadClass(name)
+        }
+
+        val found = findLoadedClass(name)
+        if (found != null) {
+          return found
+        }
+
+        val classFileName = name.replaceAll("\\.", "/") + ".class"
+        val in = currentLoader.getResourceAsStream(classFileName)
+        if (in != null) {
+          val bytes = IOUtils.toByteArray(in)
+          return defineClass(name, bytes, 0, bytes.length)
+        }
+
+        throw new ClassNotFoundException(name)
+      }
+    }
+
+    Utils.withContextClassLoader(noHive) {
+      val test = noHive.loadClass(NoHiveTest.getClass.getName().stripSuffix("$"))
+      test.getMethod("runTest").invoke(null)
+    }
+  }
+}
+
+/** Test code for SPARK-23209 to avoid using too much reflection above. */
+private object NoHiveTest {
+
+  def runTest(): Unit = {
+    try {
+      val manager = new HadoopDelegationTokenManager(new SparkConf(), new Configuration(), null)
+      require(!manager.isProviderLoaded("hive"))
```

Comment:
I think this should be extended to make sure the hdfs and hbase providers are loaded. I'm not sure what's the `ServiceLoader` behavior when an extension fails to load.

It would be good to have that test in core, btw. e.g. have a test provider that throws an exception during initialization and make sure it doesn't affect other providers.

Reply:
> I think this should be extended to make sure the hdfs and hbase providers are loaded.

I was also thinking about to add them but was hesitant to introduce inter-project tests. The more I'm considering the more I think you're right. Added.

> It would be good to have that test in core, btw. e.g. have a test provider that throws an exception during initialization and make sure it doesn't affect other providers.

Added.